File: happy-habitat-frontend/src/app/services/charges.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, catchError, throwError } from 'rxjs';
import { environment } from '../../environments/environment';
import { LoggerService } from './logger.service';
import { ErrorService } from './error.service';

export interface CargoComunidadDto {
  id: string;
  contratoId: string;
  comunidadId: string;
  montoCargo: number;
  fechaDePago: string;
  montoRecargos: number;
  estatus: string; // No vencido, vencido, pagado, pago parcial
  notas?: string | null;
  isActive: boolean;
  createdAt: string;
  updatedAt?: string | null;
}

export interface PagoComunidadDto {
  id: string;
  montoPago: number;
  formaDePago: string;
  fechaDePago: string;
  isActive: boolean;
  createdAt: string;
  updatedAt?: string | null;
  updatedByUserId?: string | null;
  pagoCargos?: PagoCargoComunidadDto[];
}

export interface PagoCargoComunidadDto {
  id: string;
  pagoComunidadId: string;
  cargosComunidadId: string;
  montoAplicado: number;
  createdAt: string;
}

export interface StatementDto {
  cargos: CargoComunidadDto[];
  pagos: PagoComunidadDto[];
}

@Injectable({
  providedIn: 'root'
})
export class ChargesService {
  private http = inject(HttpClient);
  private logger = inject(LoggerService);
  private errorService = inject(ErrorService);

  private readonly API_URL = `${environment.apiUrl}/charges`;

  /**
   * Obtiene el estado de cuenta (cargos y pagos) por contratoId
   */
  getStatementByContratoId(contratoId: string): Observable<StatementDto> {
    this.logger.debug(`Fetching statement for contract: ${contratoId}`, 'ChargesService');
    
    return this.http.get<StatementDto>(`${this.API_URL}/contrato/${contratoId}`).pipe(
      catchError((error) => {
        this.logger.error(`Error fetching statement for contract ${contratoId}`, error, 'ChargesService');
        this.errorService.handleError(error);
        return throwError(() => error);
      })
    );
  }

  /**
   * Obtiene el estado de cuenta (cargos y pagos) por comunidadId
   */
  getStatementByComunidadId(comunidadId: string): Observable<StatementDto> {
    this.logger.debug(`Fetching statement for community: ${comunidadId}`, 'ChargesService');
    
    return this.http.get<StatementDto>(`${this.API_URL}/comunidad/${comunidadId}`).pipe(
      catchError((error) => {
        this.logger.error(`Error fetching statement for community ${comunidadId}`, error, 'ChargesService');
        this.errorService.handleError(error);
        return throwError(() => error);
      })
    );
  }

  /**
   * Obtiene todos los cargos de una comunidad
   */
  getCargosByComunidadId(comunidadId: string): Observable<CargoComunidadDto[]> {
    this.logger.debug(`Fetching charges for community: ${comunidadId}`, 'ChargesService');
    
    return this.http.get<CargoComunidadDto[]>(`${this.API_URL}/comunidad/${comunidadId}/cargos`).pipe(
      catchError((error) => {
        this.logger.error(`Error fetching charges for community ${comunidadId}`, error, 'ChargesService');
        this.errorService.handleError(error);
        return throwError(() => error);
      })
    );
  }

  /**
   * Obtiene todos los cargos de un contrato
   */
  getCargosByContratoId(contratoId: string): Observable<CargoComunidadDto[]> {
    this.logger.debug(`Fetching charges for contract: ${contratoId}`, 'ChargesService');
    
    return this.http.get<CargoComunidadDto[]>(`${this.API_URL}/contrato/${contratoId}/cargos`).pipe(
      catchError((error) => {
        this.logger.error(`Error fetching charges for contract ${contratoId}`, error, 'ChargesService');
        this.errorService.handleError(error);
        return throwError(() => error);
      })
    );
  }
}

File: happy-habitat-backend/HappyHabitat.Infrastructure/Services/ChargesService.cs
using Microsoft.EntityFrameworkCore;
using HappyHabitat.Application.DTOs;
using HappyHabitat.Application.Interfaces;
using HappyHabitat.Domain.Entities;
using HappyHabitat.Infrastructure.Data;

namespace HappyHabitat.Infrastructure.Services;

public class ChargesService : IChargesService
{
    private readonly ApplicationDbContext _context;

    public ChargesService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task<StatementDto> GetStatementByContratoIdAsync(Guid contratoId)
    {
        // Obtener todos los contratos relacionados (puede haber múltiples contratos)
        var contratos = await _context.Contratos
            .Where(c => c.Id == contratoId && c.IsActive)
            .Select(c => c.Id)
            .ToListAsync();

        if (!contratos.Any())
        {
            return new StatementDto { Cargos = new List<CargoComunidadDto>(), Pagos = new List<PagoComunidadDto>() };
        }

        // Obtener cargos de todos los contratos relacionados
        var cargos = await _context.CargosComunidad
            .Where(c => contratos.Contains(c.ContratoId) && c.IsActive)
            .OrderBy(c => c.FechaDePago)
            .ToListAsync();

        // Obtener IDs de cargos
        var cargoIds = cargos.Select(c => c.Id).ToList();

        // Obtener pagos relacionados a estos cargos
        var pagoCargos = await _context.PagoCargoComunidad
            .Where(pc => cargoIds.Contains(pc.CargosComunidadId))
            .Include(pc => pc.PagoComunidad)
            .ToListAsync();

        var pagoIds = pagoCargos.Select(pc => pc.PagoComunidadId).Distinct().ToList();

        var pagos = await _context.PagoComunidad
            .Where(p => pagoIds.Contains(p.Id) && p.IsActive)
            .Include(p => p.PagoCargos)
            .OrderBy(p => p.FechaDePago)
            .ToListAsync();

        return new StatementDto
        {
            Cargos = cargos.Select(MapToCargoComunidadDto).ToList(),
            Pagos = pagos.Select(p => MapToPagoComunidadDto(p, pagoCargos.Where(pc => pc.PagoComunidadId == p.Id).ToList())).ToList()
        };
    }

    public async Task<StatementDto> GetStatementByComunidadIdAsync(Guid comunidadId)
    {
        // Obtener cargos de la comunidad
        var cargos = await _context.CargosComunidad
            .Where(c => c.ComunidadId == comunidadId && c.IsActive)
            .OrderBy(c => c.FechaDePago)
            .ToListAsync();

        // Obtener IDs de cargos
        var cargoIds = cargos.Select(c => c.Id).ToList();

        // Obtener pagos relacionados a estos cargos
        var pagoCargos = await _context.PagoCargoComunidad
            .Where(pc => cargoIds.Contains(pc.CargosComunidadId))
            .Include(pc => pc.PagoComunidad)
            .ToListAsync();

        var pagoIds = pagoCargos.Select(pc => pc.PagoComunidadId).Distinct().ToList();

        var pagos = await _context.PagoComunidad
            .Where(p => pagoIds.Contains(p.Id) && p.IsActive)
            .Include(p => p.PagoCargos)
            .OrderBy(p => p.FechaDePago)
            .ToListAsync();

        return new StatementDto
        {
            Cargos = cargos.Select(MapToCargoComunidadDto).ToList(),
            Pagos = pagos.Select(p => MapToPagoComunidadDto(p, pagoCargos.Where(pc => pc.PagoComunidadId == p.Id).ToList())).ToList()
        };
    }

    public async Task<IEnumerable<CargoComunidadDto>> GetCargosByComunidadIdAsync(Guid comunidadId)
    {
        var cargos = await _context.CargosComunidad
            .Where(c => c.ComunidadId == comunidadId && c.IsActive)
            .OrderBy(c => c.FechaDePago)
            .ToListAsync();

        return cargos.Select(MapToCargoComunidadDto);
    }

    public async Task<IEnumerable<CargoComunidadDto>> GetCargosByContratoIdAsync(Guid contratoId)
    {
        var cargos = await _context.CargosComunidad
            .Where(c => c.ContratoId == contratoId && c.IsActive)
            .OrderBy(c => c.FechaDePago)
            .ToListAsync();

        return cargos.Select(MapToCargoComunidadDto);
    }

    private static CargoComunidadDto MapToCargoComunidadDto(CargosComunidad cargo)
    {
        return new CargoComunidadDto
        {
            Id = cargo.Id.ToString(),
            ContratoId = cargo.ContratoId.ToString(),
            ComunidadId = cargo.ComunidadId.ToString(),
            MontoCargo = cargo.MontoCargo,
            FechaDePago = cargo.FechaDePago,
            MontoRecargos = cargo.MontoRecargos,
            Estatus = cargo.Estatus,
            Notas = cargo.Notas,
            IsActive = cargo.IsActive,
            CreatedAt = cargo.CreatedAt.ToString("O"),
            UpdatedAt = cargo.UpdatedAt?.ToString("O")
        };
    }

    private static PagoComunidadDto MapToPagoComunidadDto(PagoComunidad pago, List<PagoCargoComunidad> pagoCargos)
    {
        return new PagoComunidadDto
        {
            Id = pago.Id.ToString(),
            MontoPago = pago.MontoPago,
            FormaDePago = pago.FormaDePago,
            FechaDePago = pago.FechaDePago,
            IsActive = pago.IsActive,
            CreatedAt = pago.CreatedAt.ToString("O"),
            UpdatedAt = pago.UpdatedAt?.ToString("O"),
            UpdatedByUserId = pago.UpdatedByUserId?.ToString(),
            PagoCargos = pagoCargos.Select(pc => new PagoCargoComunidadDto
            {
                Id = pc.Id.ToString(),
                PagoComunidadId = pc.PagoComunidadId.ToString(),
                CargosComunidadId = pc.CargosComunidadId.ToString(),
                MontoAplicado = pc.MontoAplicado,
                CreatedAt = pc.CreatedAt.ToString("O")
            }).ToList()
        };
    }
}
